File: packages/ogdf.js/src/module/src/Ranking.cpp
#include "../main.h"

#include <ogdf/layered/OptimalRanking.h>
#include <ogdf/layered/LongestPathRanking.h>
#include <ogdf/layered/CoffmanGrahamRanking.h>

EM_PORT_API(OptimalRanking *)
RankingModule_OptimalRanking(bool separateMultiEdges, AcyclicSubgraphModule *subgraph)
{
    OptimalRanking *optimalRaking = new OptimalRanking();
    optimalRaking->separateMultiEdges(separateMultiEdges);
    optimalRaking->setSubgraph(subgraph);

    return optimalRaking;
}

EM_PORT_API(CoffmanGrahamRanking *)
RankingModule_CoffmanGrahamRanking(int width, AcyclicSubgraphModule *subgraph)
{
    CoffmanGrahamRanking *coffmanGrahamRanking = new CoffmanGrahamRanking();
    coffmanGrahamRanking->width(width);
    coffmanGrahamRanking->setSubgraph(subgraph);
    return coffmanGrahamRanking;
}

EM_PORT_API(LongestPathRanking *)
RankingModule_LongestPathRanking(bool alignBaseClasses, bool alignSiblings, bool optimizeEdgeLength, bool separateDeg0Layer, bool separateMultiEdges, AcyclicSubgraphModule *subgraph)
{
    LongestPathRanking *longestPathRanking = new LongestPathRanking();
    longestPathRanking->alignBaseClasses(alignBaseClasses);
    longestPathRanking->alignSiblings(alignSiblings);
    longestPathRanking->optimizeEdgeLength(optimizeEdgeLength);
    longestPathRanking->separateDeg0Layer(separateDeg0Layer);
    longestPathRanking->separateMultiEdges(separateMultiEdges);
    longestPathRanking->setSubgraph(subgraph);
    return longestPathRanking;
}

File: packages/ogdf.js/src/module/config/Ranking.js
import createModule from '../module-helper'
import AcyclicSubgraphModule from './AcyclicSubgraph'
import { PARAMETER_TYPE } from '../../utils/parameter-type'
const NAME = 'RankingModule'

const MODULE_DIRECTORY = {
    CoffmanGrahamRanking: {
        width: {
            type: PARAMETER_TYPE.INT,
            range: [0, Infinity],
            default: 3
        },
        subgraphType: {
            type: PARAMETER_TYPE.MODULE,
            module: AcyclicSubgraphModule,
            default: AcyclicSubgraphModule.DfsAcyclicSubgraph
        }
    },
    LongestPathRanking: {
        alignBaseClasses: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        },
        alignSiblings: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: false
        },
        optimizeEdgeLength: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        },
        separateDeg0Layer: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        },
        separateMultiEdges: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        },
        subgraphType: {
            type: PARAMETER_TYPE.MODULE,
            module: AcyclicSubgraphModule,
            default: AcyclicSubgraphModule.DfsAcyclicSubgraph
        }
    },
    OptimalRanking: {
        separateMultiEdges: {
            type: PARAMETER_TYPE.BOOL,
            range: [true, false],
            default: true
        },
        subgraphType: {
            type: PARAMETER_TYPE.MODULE,
            module: AcyclicSubgraphModule,
            default: AcyclicSubgraphModule.DfsAcyclicSubgraph
        }
    }
}

const RankingModule = createModule(NAME, MODULE_DIRECTORY)
export default RankingModule
